sklearn/manifold/_utils.pyx

```diff
@@ -109,3 +109,111 @@ cpdef np.ndarray[np.float_t, ndim=2] _binary_search_perplexity(
         print("[t-SNE] Mean sigma: %f"
               % np.mean(math.sqrt(n_samples / beta_sum)))
     return P
+
+
+@cython.boundscheck(False)
+cpdef np.ndarray[np.float_t, ndim=2] _binary_search_perplexity_nn(
```

Comment:
There are only 6 lines that have been changed in comparison to `_binary_search_perplexity`. It should be possible to combine both functions.